src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -519,19 +549,24 @@ private void removeFromInFlightRequests(long executionDd) {
    * @param response    the failed response from ES
    * @param executionId the execution id of the request associated with the response
    */
-  private synchronized void reportBadRecord(BulkItemResponse response, long executionId) {
+  private synchronized void reportBadRecord(SinkRecordAndOffset record,
+                                            BulkItemResponse response,
+                                            long executionId) {
     if (reporter != null) {
-      List<SinkRecord> sinkRecords = inFlightRequests.getOrDefault(executionId, new ArrayList<>());
-      SinkRecord original = sinkRecords.size() > response.getItemId()
+      List<SinkRecordAndOffset> sinkRecords =
+          inFlightRequests.getOrDefault(executionId, new ArrayList<>());
+      SinkRecordAndOffset original = sinkRecords.size() > response.getItemId()
           ? sinkRecords.get(response.getItemId())
           : null;
       if (original != null) {
-        reporter.report(
-            original,
+        Future<Void> reportFuture = reporter.report(
+            original.sinkRecord,
             new ReportingException("Indexing failed: " + response.getFailureMessage())
         );
+        record.offsetState.markProcessed(reportFuture::isDone);
       }
     }
+    record.offsetState.markProcessed();
```

Comment:
Doesn't this line end up always getting called even if we went into `record.offsetState.markProcessed(reportFuture::isDone);` ? You guard against the duplicate call in the `markProcessed` method, but the code flow is still awkward.